internal/command/diff_inputs.go

```diff
@@ -0,0 +1,390 @@
+		"Do not show anything, exit with 0 if the inputs are the same, 2 if they are different or 1 if an error occurs")
+
+	cmd.Flags().StringVar(&cmd.inputStr, "input-str", "",
+		"include a string as input")
+
+	return &cmd
+}
+
+// diffArgs returns an error in the following scenarios:
+// - there is less than or greater than 2 args specified
+// - the <APP-NAME> or <TASK-NAME> is a wildcard character (*)
+// - either arg is not in the format <APP-NAME>.<TASK-NAME> or a numeric value
+func diffArgs() cobra.PositionalArgs {
+	return func(cmd *cobra.Command, args []string) error {
+		if len(args) != 2 {
+			return fmt.Errorf("accepts 2 args, received %d", len(args))
+		}
+
+		containsWildCardPattern := "^\\*\\..+$|^.+\\.\\*$"
+		isNumericPattern := "^\\d+$"
+		containsTaskPattern := "^.+\\..+$"
+
+		for _, arg := range args {
+			containsWildCard, err := regexp.MatchString(containsWildCardPattern, arg)
+
+			if err != nil {
+				return err
+			}
+
+			if containsWildCard {
+				return fmt.Errorf("%s contains a wild card character, wild card characters are not allowed", arg)
+			}
+
+			isValid, err := regexp.MatchString(fmt.Sprintf("%s|%s", isNumericPattern, containsTaskPattern), arg)
+
+			if err != nil {
+				return err
+			}
+
+			if !isValid {
+				return fmt.Errorf("%s does not specify a task or task-run ID", arg)
+			}
+		}
+
+		return nil
+	}
+}
+
+func (c *diffInputsCmd) run(cmd *cobra.Command, args []string) {
+	if args[0] == args[1] {
+		exitOnErr(fmt.Errorf("%s and %s refer to the same task-run", args[0], args[1]))
+	}
+
+	app1, task1, runID1 := c.parseDiffSpec(args[0])
+	inputs1, run1, err := c.getTaskRunInputs(app1, task1, runID1)
+	if err != nil {
+		exitOnErr(err)
+	}
+
+	app2, task2, runID2 := c.parseDiffSpec(args[1])
+	inputs2, run2, err := c.getTaskRunInputs(app2, task2, runID2)
+	if err != nil {
+		exitOnErr(err)
+	}
+
+	if run1 != nil && run2 != nil {
+		if run1.ID == run2.ID {
+			exitOnErr(fmt.Errorf("%s and %s refer to the same task-run", args[0], args[1]))
+		}
+	}
+
+	if !c.quiet || c.csv {
+		c.printOutput(inputs1.Inputs(), inputs2.Inputs())
+	}
+
+	app1Digest := getDigest(inputs1)
+	app2Digest := getDigest(inputs2)
+
+	if app1Digest == app2Digest {
+		exitFunc(0)
+	}
+
+	exitFunc(2)
+}
+
+// parseDiffSpec splits an argument into the app, task and runID components.
+// It relies on the fact that the arguments have already been validated
+// in the cobra.Command
+func (c *diffInputsCmd) parseDiffSpec(s string) (app, task, runID string) {
+	if strings.Contains(s, ".") {
+		spl := strings.Split(s, ".")
+		app = spl[0]
+		task = spl[1]
+
+		if strings.Contains(task, "^") {
+			caretIndex := strings.Index(task, "^")
+			runID = subStr(task, caretIndex, len(task))
+			task = subStr(task, 0, caretIndex)
+		}
+
+		return app, task, runID
+	}
+
+	return "", "", s
+}
+
+func (c *diffInputsCmd) getTaskRunInputs(app, task, runID string) (*baur.Inputs, *storage.TaskRunWithID, error) {
```

Comment:
Currently `getTaskRunInputs()` is called 2x when the command is invoked.
Both times `mustFindRepository()` is called, `mustArgToTask()` can be called twice depending on the commandline parameters.
This means everytime `getTaskRunInputs()` is called, the repository has to be located in the filesystem, the baur config has to be loaded and passed, the application config files have to be found, they have to be read and parsed.

Doing these operations multiple times is a lot of unnecessary overhead which can be prevented:

- call `mustFindRepository()` only 1x in the parent function
- if both commandline parameters refer to tasks in the repository (`<APPNAME>.<TASK-NAME>`), call the `mustArgToTask()` function only 1x and pass both parameters (`mustArgToTask(param1, param2)`)
- would be good to call `mustNewCompatibleStorage(repo)` also only 1x in the parent function, does not have such a  huge overhead but it's also unnecessary to create the psql client 2x